README.txt

Contact email: <imarkov@umich.edu>

To configure UCLApack, type 

script install.log
perl config 

and answer a few questions (make sure you use PERL 5+ by first typing perl -v).
When you see command prompt type 

exit

If something goes wrong, send install.log along with your bug report.

It is recommended that you let the installation program build all the 
libraries for you and then let it test them (i.e. answer yes to those 
questions) --- just lay back and watch the screen for a while. No compiler 
errors / warnings should appear. Test results should not DIFFer from 
precomputed output files.

---------------------

  Use notes:
         0. You need 300-400Mb of disk space available to install 
            UCLApack and run regression tests.
         1. Currently supported platforms are Solaris/CC5.1, Solaris/g++2.95.2
            and Linux/g++2.95.2 (you can get g++ from sunfreeware.com).
            Solaris version must be 2.7 or higher. Earlier compilers will not
            work; Solaris 2.5 or earlier will not work.
            CC -V, g++ -v and uname -a should tell you the version numbers
            of compilers and systems. Please include those numbers into all
            bugs reports that you send to developers.
            g++ 2.7.2, CC 4.1, CC4.2 and CC5.0 compilers are not and 
            will not be supported. The C++ code compiled with MSVC++ 5.0 
            and 6.0, but we currently do not support automated installation
            for MS Windows.
         2. Installation program requires PERL 5+
            If you have perl4, you are in the Stone Age ---
            please get perl5 or higher (or try perl5 on your system,
            if it works alias perl to perl5 before running this script).
         3. The LEF/DEF parser that may be enclosed with distribution
            requires flex and bison. Those come preinstalled on Linux, but
            may need to be downloaded from sunfreeware.com for Solaris.
         4. The latest g++ RPMs for RedHat Linux can be downloaded from
            ftp://ftp.varesearch.com/pub/support/hjl/gcc/gcc-2.95.x/
         5. In order to link the package ParserLEFDEF and all packages 
            that use it, the library libfl.a must be available to the linker
            using its default search path or in the lib/ subdirectory of
            UCLApack (at the moment when linking starts). libfl.a is a part of
            GNU flex (similar to lex) and comes with most Linux systems.
            On Solaris, if you already have that library, you can put a 
            symbolic link into one of those locations rather than copy 
            the actual library. 
            Successful installations were reported for GNU bison 1.25 and GNU 
            flex 2.4.5, 2.5.3 or 2.5.4 (all work). We provide a binary libfl.a
            for download that is compiled from flex 2.5.4. 



DB/CHANGES

  CHANGES TO DB

1.3  970729 ilm  made dbMasterCell::_site (see dbCell.h)
                          a reference to dbSite (was dbSite)

                 changed dbMasterPin::_offset to dbMasterPin::_center

                 commented out dbPin::_name

                 created  DBpostProc.cxx and started DB::postProcessor()

                 DB::postProcessor() is called by DB::DB() before isConsistent()

                 dbCellType::dbCellType() now abkfatals() . dbCellType will be

                 set by the postProcessor(), so we'll need the ctor then.

                 added a default constructor for dbCellType as well as 
                             MacroUNITIALIZED constrant and bool isInitialized()

                 dbCell ctor does not take cellType anymore (cellType will
                             be set in the postProcessor)

1.4  970729 ilm  moved moved enums (used for types in dbContext.h) inside
                       respective classes simplifying the enum names and 
                       the enumed names

                 dbLayer::dbLayer now initializes all members

                 changed dbLayer::_capmult => dbLayer::capMult
                         dbLayer::_edgecap => dbLayer::_edgeCap
                          (access methods too)

                 deleted dbLayer::_startCoord. dbLayer::_numTracks will    
                         eventually be set by DB::postProcessor(), but 
                         parsers can put someinfo there for postprocessor 
                         (we will standardize what exactly after the first
                          parser works ;-)
 
       ilm/aec   default ctor dbNetlist now initialized the 3 aggregate members

1.5 970730 wyd   changed dbMasterPin back to 1.2 regarding _name.

		 changed dbPin accessor getName() to getMasterPinName().

		 taken away _numTracks from dbLayer constructor.

		 added MakeFriendsWithParsers to class DB.

		 corrected typo error in dbDefines.h (parser -> Parser)
		 also typo in dbContext.h: dbLayer::Type

           ilm   added a missing const next to dbPin::getMasterPinName()

                 added a missing dbLayout::getNumRows()

1.6 970731 ilm   Still reworking Makefile for db to link with two parsers

	         Added methods to DB for LEFDEFq

		 Split DB.cxx into pathDelims.h, DB.cxx, DBreadLEFDEF.cxx
			  DBreadLEFDEFq.cxx

                 MSVC++ portability change A::Type -> Type (when inside A)

1.7 970731  ilm  Finally figured out a reasonable Makefile

1.8 970803  aec  postprocessor fills cells' nets and net's cells.

1.9 970804  ilm  now LGC file name is passed to the postprocessor only
            
                 added LEFDEFqLGC file format
 
                 added a hook for saveDEF(const char *auxFileName) const

                 added a hook for dbSpatial::reorderCells(const Permutation&)

                 changed what looked like a typo in class dbMasterCellType
                              in dbTypes.h (enum LClass into enum MClass)

                 made enums public in all type classes and added "direct"
                                      constructors (as requested by Mike)

                 dbLayer ctor now sets _height; old ctor is still there 
                                  for compatibility, but will be deleted soon

                 parsers are not friends to dbLayer since ctor does everything

                 added unified checks  (not finished)

                 abkassert around consistency check now functional

                 added three accessors to dbMasterCell

	   aec	 changed dbCell constructor to set inDegree/outDegree
		
		 added postprocessing to set instance cell types
	
		 added _numIOPads, _numAllPads and _coreCellArea to dbNetlist
			they are set by the postProcessor
	
		 added private function dbMasterCellType::setCoreClass(char*)
		 added checking of all 'nums' and total areas in consist. chk.

		 
	  		 
1.10 970805 aec  dbNetlist is now a friend to dbMasterCell for
			postProcessing of LGC files.

		 added setCoreCell in dbMasterCellType.

  		 added processing of LGC files 

1.11 970806 aec  change LGC processing..entries for non-core cells are now
			just ignored.

     970806 ilm  the old dbLayer ctor is gone
     
     970806 ilm  consistency checker now prints the types and numbers of
                 ctraints

1.12 970807 ilm  added row/site/layer accessors to class dbContext
     970807 ilm  added cell and pin "by index" accessors to class dbNet
     970807 aec  changed cerr's to abkwarn's in conchk.
		 added checking that each cell's nets are in design's vector
		 added checking that each net's cells are in design's vector
		 added isConnected() to dbPin
		 net's of degree0 are now just a msg, not an error
		 special nets are removed from each cell's vector of nets

1.13         (aec)   ??

1.14 970808  mro Added new dbMasterCell ctor using enum types and obstacles
     970808  mro added pragmas to get rid of warnings in MSVC++
     970808  mro changed "UnknownCC" to "UnknownCoreClass"

1.15         aec changes to sorting in dbPostProc.h.  Added sorting of all
			cells name, then nets by cell indexes, and cells
			in pads first order.	
     970812  ilm changed Makefile to use the "alpha" versions of libparserLEFDEF
                      (i.e. Mike's parser)

1.16 970812  ilm corrected a bug in the AUX file reader 
                   (missing 0 char at the end of AUX file directory name)
             ilm renamed the AUX directory into EXAMPLES directory
	     aec added numCCs and numSCs to dbNetlist.
             ilm calls to Mike's parser commented out to avoid problems
                     with Cadence license manager

1.17 970812  ilm fixed two bugs in the Constraints package
             wyd fixed a bug in parserLEFDEFq

1.18 970813   
         aec/pat added dbEasyAccess
             mro added double() around _ranIntWidth in four places

1.18.1 970813  
             mro moved struct equnsigned in dbEasyAcess.cxx to file scope

DB/dbNet.h
/**************************************************************************
***    
*** Copyright (c) 1995-2000 Regents of the University of California,
***               Andrew E. Caldwell, Andrew B. Kahng and Igor L. Markov
*** Copyright (c) 2000-2002 Regents of the University of Michigan,
***               Saurabh N. Adya and Igor L. Markov
***
***  Contact author(s): abk@cs.ucsd.edu, imarkov@umich.edu
***  Original Affiliation:   UCLA, Computer Science Department,
***                          Los Angeles, CA 90095-1596 USA
***
***  Permission is hereby granted, free of charge, to any person obtaining 
***  a copy of this software and associated documentation files (the
***  "Software"), to deal in the Software without restriction, including
***  without limitation 
***  the rights to use, copy, modify, merge, publish, distribute, sublicense, 
***  and/or sell copies of the Software, and to permit persons to whom the 
***  Software is furnished to do so, subject to the following conditions:
***
***  The above copyright notice and this permission notice shall be included
***  in all copies or substantial portions of the Software.
***
*** THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, 
*** EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
*** OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
*** IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
*** CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT
*** OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR
*** THE USE OR OTHER DEALINGS IN THE SOFTWARE.
***
***
***************************************************************************/











//	CHANGES
// 970806 aec removed 'const' from _index, postProcessor changes it when
//		it re-orders nets.
//	      made dbNetlist a friend to dbNet
// 970807 ilm added cell and pin "by index" accessors to class dbNet
// 970814 aec added friend class DB
// 970815 aec added const to 'loopCheck' functions
// 970817 ilm added "compare" classes from dbSortComp.h
// 970819 aec added CompareNetsByDegree
// 970825 ilm improved diagnostics for nested forall macros
// 970905 aec added dbNet::_numSources
// 970906 aec changed dbNet::_numSources to _numOut, added _numInOut
//		added sourcesBegin/End, sourceSinksBegin/End, sinksBegin/End
//		iterators for cells.
// 971130 ilm added "prohibiting" copy-ctor and op= for dbNet
// 971209 mro added snetOffset to distinguish special-net indices
//            from ordinary ones, isSpecial() inspector, and CompareNetsByName
// 971214 ilm publicized ("nationalized") cells/pins in dbNet


#ifndef __DB_NET_H_
#define __DB_NET_H_

#include <algorithm>
#include "dbDefines.h"
#include "dbTypes.h"
#include "dbPin.h"
#include "dbCell.h"
#include "dbNet.h"
#include "dbContext.h"

const unsigned snetOffset = 1000000000; //special-net indices start here

class dbNet
{
    MakeFriendsWithParsers
    friend class dbNetlist;
    friend class DB;

    char *      	_name;
    unsigned   		_index;
    const dbNetType  	_type;        // regular, timing, clock
 public:
    ctainerCellsLocal  	_cells;       // source cell(s) come<s> first
    ctainerPinsLocal   	_pins;
    ctainerLayerGeoms  	_wiring;      // should be empty except for special
                                      // nets.
    unsigned		_numOut;
    unsigned		_numInOut;

    bool                _hasLogicalPin; 
    double              _weight;      //weight of the net. 

 private:    
    static char      _forAllCellsFlag;
    static char      _forAllPinsFlag;   
    
    // net weights should be taken care of by placer/elsewhere

        // this is what you can't do  --
        // default copy-ctor and assignment are wrong

        dbNet(const dbNet& from):_type(from._type)
                 { abkfatal(0,"Can't copy-construct dbNet"); }
        void operator=(const dbNet&)
                 { abkfatal(0,"Can't assign dbNet");}

    
    public:
        dbNet(const char * name, unsigned idx, dbNetType type,
              const ctainerPinsLocal& pins):
	      _index(idx),_type(type),_pins(pins),_hasLogicalPin(false),
	      _weight(1)
              {  /* Extrapolate vector<dbCell*> cells from pins */
              /* Go through each pins and set the cell ptr */
              /* This should be verified as part of a future integrity check */
              _name = new char[strlen(name)+1];strcpy(_name,name);
              }
        ~dbNet();
        unsigned  getNumCells() const { return _cells.size(); }
        unsigned  getNumPins()  const { return _pins.size() ; }
        unsigned  getDegree()   const { return _cells.size(); }
        unsigned  getIndex()    const { return _index; }
        dbNetType getType()     const { return _type; }
        const char* getName()   const { return _name; }
 	unsigned  getNumOut()   const { return _numOut; }
	unsigned  getNumInOut() const { return _numInOut; }
	unsigned  getNumIn()	const { return _cells.size() - 
					(_numOut + _numInOut); }
	double    getWeight()   const { return _weight; }

        unsigned   hasLogicalPin() const {return _hasLogicalPin;}
          
        itCellLocal cellsBegin() const { return _cells.begin(); }
        itCellLocal cellsEnd()   const { return _cells.end();   }

	itCellLocal sourcesBegin() const { return _cells.begin(); }
	itCellLocal sourcesEnd()   const { return _cells.begin() + _numOut; }

	itCellLocal sourceSinksBegin() const 
		{ return _cells.begin() + _numOut; }
	itCellLocal sourceSinksEnd()   const 
		{ return _cells.begin() + _numOut + _numInOut; }

	itCellLocal sinksBegin() const 
		{ return _cells.begin() + _numOut + _numInOut; }
	itCellLocal sinksEnd()   const 
		{ return _cells.end(); }

        const dbCell&  getCellByIdx(unsigned i) const
        {
          abkassert(i < _cells.size(),"Cell index out of range");
          return *(_cells[i]);
        }

        itPinLocal pinsBegin()  const { return _pins.begin();  }
        itPinLocal pinsEnd()    const { return _pins.end();    }

        const dbPin&  getPinByIdx(unsigned i) const
        {
          abkassert(i < _pins.size(),"Pin index out of range");
          return *(_pins[i]);
        }

        itLayerGeom wiringBegin() const { return _wiring.begin();}
        itLayerGeom wiringEnd()  const { return _wiring.end();  }

        const dbLayerGeom& getWiringLayerGeom(unsigned i) const
              {return *(_wiring[i]);}

        unsigned getNumWiringLayerGeoms() const {return _wiring.size();}
          
        void loopCellsBegin() const;
        void loopCellsEnd(itCellLocal itr) const;
        void loopPinsBegin() const;
        void loopPinsEnd(itPinLocal itr) const;
          
          
        const dbPin *commonPin(const dbCell* cell) const;
        bool   commonPinIsCellInput (const dbCell* cell) const;
        bool   commonPinIsCellOutput(const dbCell* cell) const;
          
        double getInputCap(const dbCell* cell) const;
        double getDriverResist(const dbCell* cell) const;

        bool isSpecial() const {return _index >= snetOffset;} //CAREFUL:
                                                              //works only
                                                              //*after*
                                                              //postprocessing

    private:		
	void induceCellsFromPins();	
	void sortCellsByIndex();
	void sortCellsByDir();
        void removeDupPins();
          
};

class CompareNetsByCellIndex    //comp. class for sorting the netlist's nets
{
   public:
    CompareNetsByCellIndex(){}

    bool operator()( dbNet* n1, dbNet* n2) const;
};

class CompareNetsByDegree
{
    public:
     CompareNetsByDegree(){}

     bool operator()( dbNet* n1, dbNet* n2) const;
};

class CompareNetsByName
{
    public:
     CompareNetsByName(){}

     bool operator()( dbNet* n1, dbNet* n2) const;
};

class CompareNetsByDirection
{
    const dbCell& _cell;
  public:
    CompareNetsByDirection(const dbCell& cell)
	: _cell(cell)
	{}
    bool operator()( dbNet* n1, dbNet* n2) const;
};

// =========================  IMPLEMENTATIONS ========================

inline void dbNet::sortCellsByIndex()
{
    std::sort(_cells.begin(), _cells.end(), CompareCellsByCellIndex());
}

inline void dbNet::sortCellsByDir()
{
    std::sort(_cells.begin(), _cells.end(), CompareCellsByDir(this));
}

inline void dbNet::loopCellsBegin() const
{
    abkassert(!_forAllCellsFlag,
        "ERROR: nesting of forallCells macros is invalid");
    _forAllCellsFlag = true;
}

inline void dbNet::loopCellsEnd(itCellLocal itr) const
{
    if(itr == cellsEnd())_forAllCellsFlag = false;
}


inline void dbNet::loopPinsBegin() const
{
    abkassert(!_forAllPinsFlag,
        "ERROR: nesting of forallPins macros is invalid");
    _forAllPinsFlag = true;
}

inline void dbNet::loopPinsEnd(itPinLocal itr) const
{
    if(itr == pinsEnd())_forAllPinsFlag = false;
}

#endif

